src/bigints.nim

```
@@ -956,3 +956,54 @@ iterator `..<`*(a, b: BigInt): BigInt =
   while res < b:
     yield res
     inc res
+
+func invmod*(a, modulus: BigInt): BigInt =
+  ## Compute the modular inverse of `a` modulo `modulus`. 
+  ## The return value is always in the range `[1, modulus-1]`
+  # Euclide's method
+  if a == 0:
+    raise newException(DivByZeroDefect, "0 has no modular inverse")
+  var
+    r0 = a
+    r1 = modulus
+    u = one
+    u1 = zero
+    q, rt, ut : BigInt
+  while r1 > 0:
+    q = r0 div r1
+    rt = r0
+    ut = u
+    r0 = r1
+    u = u1
+    r1 = rt - q * r1
+    u1 = ut - q * u1
+  if r0 != one:
+    raise newException(ValueError, $a & " has no modular inverse")
+  result = ((u mod modulus) + modulus) mod modulus
+
+func powmod*(base, exponent, modulus: BigInt): BigInt =
+  ## Compute modular exponentation of `base` with power `exponent` modulo `modulus`.
+  ## The return value is always in the range `[0, modulus-1]`.
+  if modulus < 0:
+    raise newException(ValueError, "modulus must be strictly positive")
+  elif modulus.isZero:
+    raise newException(DivByZeroDefect, "modulus must be nonzero")
+  elif modulus == 1:
+    return zero
+  elif exponent < 0:
+    if base.isZero:
+      return zero
```

I think there's nothing speaking against moving this check up a level, since we have it in the code anyway. Besides, is this really the common behavior? For normal integers, `0 ^ exponent` is undefined if `exponent` is negative.

---

Negative exponent for modular exponentiation is really common. `0 ^ exponent = 0` for all `exponent != 1` is a general convention. It is common to make the convention `0^{-1} = 0`, it enables to consider functions like `x -> x ^ p-1 mod p` as bijections for all elements of Z/pZ (0 included) in mathematics. But your concern is justified. `0 ^ -n = (0 ^ -1) ^ n` would be the inverse of `0` (raised to exponent `n`) for which I have raised an exception in invmod.
 It makes the code inconsistent.
Moving this check up a level means we do an additional check for the most common case (base > 0, exponent > 0), whereas we optimize an exceptional case (base = 0) that should almost never happen. base.isZero check is not totally free though (it uses a whole loop as of yet).

---

I see your point, but note that the loop in `isZero` should run at most once (if the number is correctly normalized). Feel free to mark this as resolved.

---

I will finally raise an exception for negative exponent of 0, as 0 is not invertible. I wonder if I should raise an exception inside powmod, or let invmod raise the exception (this would avoid the check).

---

I think letting `invmod` raise the exception is fine.